requests-for-comments/2024-05-07-repeatable-surveys.md

```diff
@@ -0,0 +1,75 @@
+   to repeat a survey, they'll start using it.
+
+## Context
+_What are our competitors doing, what are the technical constraints, what are customers asking for, what does the data tell us, are there external motivations (e.g. launch week, enterprise contract)?_
+
+Our competitors have already built tools that help customers track user sentiment over a time period, allowing them to repeat surveys and gain insights into changing user behavior.
+
+Our current workarounds of `just repeat the survey manually` might get them to repeat the survey, but there's no way to gain any useful insights since the responses will end up in different surveys.
+
+### Similar tools
+1. [Pendo's custom NPS survey type](https://support.pendo.io/hc/en-us/articles/360033527151-Set-up-an-NPS-survey#h_01HGD0SKYSRB6SNKXBFJ1QG7M7)
+2. [Surveymonkey's multiple responses](https://help.surveymonkey.com/en/surveymonkey/send/allowing-multiple-responses/)
+
+
+## Design
+_What are the key user experience and technical design decisions / trade-offs?_
+
+A survey is assumed to have only one response in the system and we have filter conditions that govern **who** can see the survey.
+A user's survey response action is tracked as a boolean, i.e. we add a person property with the survey-id `$survey_response/<surveyID>` when a user submits or dismisses a review.
+
+To support this feature, we need to introduce the concept of `survey iterations` to the UI and person properties.
+
+Here's what we need to do to solve this : 
+
+### Data model changes to the surveys table
+1. Add properties to the survey data model to track 
+
+| Column name | type | description | nullable |
+|:----|:----|:----|:----|
+| `survey_frequency` | &lt;some cron indicator, could be varchar&gt; |  Frequency at which the survey should be repeated | Yes | 
+| `survey_max_iteration_count` | integer | max number of times the survey should be repeated |  Yes |
+
+This design allows our users to setup such surveys:
+
+1. `repeat this survey 4 times, once every 3 months`.
+2. They can also just leave the `survey_max_iteration_count` as zero to have an infinitely repeating survey `repeat this survey, once every 3 months`
+
+These two properties are nullable and we need to make our work backwards compatible to make existing survey responses function without issues to show NPS scores etc. 
+
+#### UI changes. 
+1. Show the user an interface to pick the repeat interval of the survey, this can be relative (every 4 months) or absolute (pick specific dates). 
+4. Show the user an interface to pick the iteration count, how many times should this survey repeat? 
+5. If a survey has an iteration count of more than 1, change the existing UI to run custom clickhouse queries to account for a survey with multiple response sets, one for each iteration. 
+
+#### posthog-js changes.
+1. To decide when to show a survey, see if a user has a survey response submitted for the current survey start date, and show the survey if they haven't.
```

Comment:
> see if a user has a survey response submitted for the current survey start date

How are you planning to query this?

Also worth noting that you'll need to put the local storage cache on a ttl / find another way around not showing the same survey repeatedly. There are some complex patterns here to handle:

1. When a user first dismisses/responds to a survey, we add something to local storage to not show this survey again.
2. Now, I'm assuming there's some mechanism like flags(?) controlling when to next show it. This will almost certainly have some delay (since we need to send properties back via ingestion, which can take some time).
3. Because of the async nature of (2), the client side needs some way to remember not to show this survey again, hence (1). But now, if we do want to re-show it, we also want to delete (1), hence some ttl there.

Reply:
This is, as you noted, going to be complicated. 
For a given survey iteration, we can't depend on the local storage having the necessary information to show/hide a survey given how long the iterations last. 
To get over this, we could associate each iteration of the survey with its own feature flag, that has a time based activation.

e.g. `survey-targeting-flag-iter-1` is only true for 90 days, until Aug 11 2024. and 
`survey-targeting-flag-iter-2`  is from Aug 11 2024 to November 9th 2024. And we create these flags on a schedule based on the survey frequency. `posthog-js` just uses this feature flag information to show/hide the dialog. And when it sends the event, it sends the `iteration_id` as a person property, from the survey object properties.